Add example with rabbitmq hub impl

aashikam opened the pull request:
## Purpose
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/1713

## Examples

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

shafreenAnfar commented:
We decided to remove the consolidater and keep the implementation simple. We need to update the image accordingly as well.

ayeshLK commented on examples/rabbitmq-hub/subscriber/subscriber_service.bal:
```diff
@@ -0,0 +1,35 @@
+import ballerina/websub;
+import ballerina/log;
+
+listener websub:Listener securedSubscriber = new(9100, host = "localhost");
```
```suggestion
listener websub:Listener securedSubscriber = new(9100);
```

ayeshLK commented on examples/rabbitmq-hub/subscriber/subscriber_service.bal:
```diff
@@ -0,0 +1,35 @@
+listener websub:Listener securedSubscriber = new(9100, host = "localhost");
+
+@websub:SubscriberServiceConfig {
+    target: ["http://localhost:9000/hub", "test"], leaseSeconds: 36000
```
```suggestion
    target: ["http://localhost:9000/hub", "test"], 
    leaseSeconds: 36000
```

ayeshLK commented on examples/rabbitmq-hub/publisher/topic_registration_client.bal:
```diff
@@ -0,0 +1,24 @@
+
+public function main() returns error? {
+    websubhub:PublisherClient websubHubClientEP = check new("http://localhost:9000/hub");
+    var registrationResponse = websubHubClientEP->registerTopic("test");
```
```suggestion
    websubhub:TopicRegistrationSuccess registrationResponse = check websubHubClientEP->registerTopic("test");
```

ayeshLK commented on examples/rabbitmq-hub/hub/start_hub.bal:
```diff
@@ -0,0 +1,213 @@
+}
+
+function syncSubscribersCache() returns error? {
+    do {
+        while true {
+            websubhub:VerifiedSubscription[]? persistedSubscribers = check getPersistedSubscribers();
+            if persistedSubscribers is websubhub:VerifiedSubscription[] {
+                refreshSubscribersCache(persistedSubscribers);
+                check startMissingSubscribers(persistedSubscribers);
+            }
+        }
+    } on fail var e {
+        return e;
+    }
```
```suggestion
    while true {
        websubhub:VerifiedSubscription[]? persistedSubscribers = check getPersistedSubscribers();
        if persistedSubscribers is websubhub:VerifiedSubscription[] {
            refreshSubscribersCache(persistedSubscribers);
            check startMissingSubscribers(persistedSubscribers);
        }
    }
```

ayeshLK commented on examples/rabbitmq-hub/hub/start_hub.bal:
```diff
@@ -0,0 +1,213 @@
+}
+
+function syncRegisteredTopicsCache() returns error? {
+    do {
+        while true {
+            websubhub:TopicRegistration[]? persistedTopics = check getPersistedTopics();
+            if persistedTopics is websubhub:TopicRegistration[] {
+                refreshTopicCache(persistedTopics);
+            }
+        }
+    } on fail var e {
+        return e;
+    }
```
```suggestion
    while true {
        websubhub:TopicRegistration[]? persistedTopics = check getPersistedTopics();
        if persistedTopics is websubhub:TopicRegistration[] {
            refreshTopicCache(persistedTopics);
        }
    }
```

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,185 @@
+    # + message - Details of the subscription
+    # + return - `error` if there is any unexpected error or else `()`
+    isolated remote function onSubscriptionIntentVerified(websubhub:VerifiedSubscription message) returns error? {
+        string groupName = util:generateGroupName(message.hubTopic, message.hubCallback);
```
```suggestion
```

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,185 @@
+    # 
+    # + message - Details of the unsubscription
+    isolated remote function onUnsubscriptionIntentVerified(websubhub:VerifiedUnsubscription message) {
+        string groupName = util:generateGroupName(message.hubTopic, message.hubCallback);
```
```suggestion
```

ayeshLK commented:
Do we need to add a `.gitignore` for the example ?

aashikam commented:
> Do we need to add a `.gitignore` for the example ?

I added because I remembered like I added it in other examples. Seems unnecessary. Removed the file.

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,183 @@
+    # Registers a `topic` in the hub.
+    # 
+    # + message - Details related to the topic-registration
+    # + headers - `http:Headers` of the original `http:Request`
```
```suggestion
```

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,183 @@
+    # De-registers a `topic` in the hub.
+    # 
+    # + message - Details related to the topic-deregistration
+    # + headers - `http:Headers` of the original `http:Request`
```
```suggestion
```

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,183 @@
+        } 
+    }
+
+    # Publishes content to the hub.
+    # 
+    # + message - Details of the published content
+    # + headers - `http:Headers` of the original `http:Request`
+    # + return - `websubhub:Acknowledgement` if publish content is successful, `websubhub:UpdateMessageError`
+    #            if publish content failed or `error` if there is any unexpected error
+    isolated remote function onUpdateMessage(websubhub:UpdateMessage message, http:Headers headers)
```
```suggestion
    # Publishes content to the hub.
    # 
    # + message - Details of the published content
    # + return - `websubhub:Acknowledgement` if publish content is successful, `websubhub:UpdateMessageError`
    #            if publish content failed or `error` if there is any unexpected error
    isolated remote function onUpdateMessage(websubhub:UpdateMessage message)
```

ayeshLK commented on examples/rabbitmq-hub/hub/hub_service.bal:
```diff
@@ -0,0 +1,183 @@
+
+import ballerina/websubhub;
+import ballerina/log;
+import ballerina/http;
```
```suggestion
```

ayeshLK commented on examples/rabbitmq-hub/README.md:
```diff
@@ -0,0 +1,76 @@
+
+# Overview
+
+This is a websub hub implementation is based on message brokers. In this case we have used RabbitMQ message broker to implement the websubhub APIs. Message brokers are a natural fit for the implementation of websub hubs as they serve a similar purpose. However, regardless of the APIs used to publish messages to the hub, websubhub APIs can be used to implement websub hubs which are fully compliant ...
```
```suggestion
The purpose of this example is to demonstrate the usage of ballerina RabbbitMQ package in a production ready application. 

This is a websub hub implementation based on message brokers. In this case hub is backed by RabbitMQ message broker. Message brokers are a natural fit for the implementation of websub hubs as they serve a similar purpose.
```

ayeshLK commented on examples/rabbitmq-hub/hub/Config.toml:
```diff
@@ -0,0 +1,31 @@
+# All the hubs must be pointed to the same RabbitMQ queue to notify websub subscription/unsubscription
+WEBSUB_SUBSCRIBERS_QUEUE = "registered-websub-subscribers"
+
+# The interval in which RabbitMQ consumers wait for new messages
+POLLING_INTERVAL = 10.0
+
+# The period in which RabbitMQ close method waits to complete
+GRACEFUL_CLOSE_PERIOD = 5.0
```
Do we need these configurations ?

ayeshLK commented:
Since `Dependencies.toml` is an auto-generated artifact shall we ignore the `Dependencies.toml` of examples by adding following line to repository `.gitignore` file.
```sh
examples/**/Dependencies.toml
```
